libcxx/docs/DesignDocs/NodiscardExtensions.rst

```diff
@@ -0,0 +1,34 @@
+======================================
+``[[nodiscard]]`` extensions in libc++
+======================================
+
+Libc++ adds ``[[nodiscard]]`` to functions in a lot more places than the
+standard does. Any applications of ``[[nodiscard]]`` that aren't required by the
+standard written as ``_LIBCPP_NODISCARD_EXT`` to make it possible to disable
+them. This can be done by defining ``_LIBCPP_DISABLE_NODISCARD_EXT``.
+
+When should ``[[nodiscard]]`` be added to functions?
+====================================================
+
```

Comment:
This would fall under "where discarding the value is most likely a misuse of the function". IMO we should diagnose these kinds of cases. In your example `any_of`, or if you want to keep identical semantics `for_each`, would be the correct algorithm to use. I think we want to drive people towards writing better code, and this helps. Also, although it's not not _super_ relevant in this case, the side effect some program is relying on may be the result of some library-internal decision that could change as the library evolves. Avoiding people writing code like that is both in their interest and ours.

Reply:
IMHO if you're using the function like that then I think being forced to explicitly cast the result to void makes it clearer for readers that you're using it in an unusual way.

Reply:
I don't feel it's poor style per se. In this case you could use `std::expected`. But I've seen similar creative usage of `std` algorithms in code bases.
IMO the cast to void is a hint to the reader something odd is going on and the algorithm is used in a different way. So I'm still happy to make these `[[nodiscard]]`.